### DISCUSSION TITLE
Frequency of Transcripts Returned by Endpointing

### DISCUSSION THREAD

**Discussion by rrichc** [deepgram/community#319] [General help] [answered]
I'm currently running the `phonecall` model on Nova with live transcription, and have some questions on endpointing and how often it returns transcripts.

With `endpointing=1000` and `interim_results: false`, I'm currently having to employ the same strategy described below, for when both endpointing and interim results are on:
> When both endpointing and interim results are enabled, do not use speech_final: true to pull a full, final transcript. If a speaker vocalizes a long utterance, Deepgram's algorithm may respond with some finalized transcripts of preliminary results (marked as is_final: true) before the end of the utterance (marked as speech_final: true) is detected.
In this case, to gather a full transcript, you must concatenate all responses marked is_final: true until you reach a response marked speech_final: true.
https://developers.deepgram.com/docs/understand-endpointing-interim-results

To me, this suggests that when interim results are off, but endpointing is on, there shouldn't be a sequence of `is_final: true` transcripts containing partial transcripts with a `speech_final: true` following to denote the end, especially if the period of silence has not elapsed the ms thresold set. Instead, I expect only a single transcript returned with `speech_final: true` fully containing the entire transcript, and only once the silence threshold as been exceeded, as described in the doc below.

> By default, endpointing is enabled and will return transcripts after detecting 10 milliseconds of silence. When endpointing is enabled, once a speaker finishes speaking, no transcripts will be sent back until the speech resumes or the required amount of silence has been detected. 
https://developers.deepgram.com/docs/endpointing 

My concern here is very similar to the discord post here: https://discord.com/channels/1108042150941294664/1129327722821337170/1129327722821337170

1. Am I to understand that regardless of whether interim results are enabled in conjunction with endpointing, Deepgram will send back parts of the finalized transcripts in `is_final: true` messages even before the threshold silence has been exceeded?

> Endpointing relies on a Voice Activity Detector

2. Does Endpointing rely on pure silence? Or the lack of what it determines to be a voice? I'm trying to figure out why occasionally when endpointing is set to higher ms, it holds back a `speech_final: true` even when no one is talking and the silence threshold has been exceeded. I've even muted myself on Twilio  to test this, and it only sends the final part of the transcript once I begin to talk again. My assumption is that it's because Twilio is a noisier connection so the threshold is not hit if based on pure silence.

I'm building a conversational AI over the phone, and to switch strategies I've decided to follow the advice in this post involving `utterance_end_ms` and `{"type":"UtteranceEnd"}`:
https://github.com/orgs/deepgram/discussions/177

3. Since I see it hasn't been documented just yet, is this considered a stable feature or are there any concerns with using it in production?

Other resources I've considered beyond the above:
https://developers.deepgram.com/docs/interim-results
https://deepgram.com/learn/building-a-conversational-ai-flow-with-deepgram

**Comment by nikolawhallon** [comment, marked as answer]
Thanks for the detailed post - this will be great to inform some docs clarifications.

> to gather a full transcript, you must concatenate all responses marked is_final: true until you reach a response marked speech_final: true

this statement is true when endpointing is turned on - when endpointing is turned off, no `speech_final` results will ever be returned, so one would still concatenate all `is_final` results (but have no indications of end-of-speech). In other words:

> I expect only a single transcript returned with speech_final: true fully containing the entire transcript

^ this does not happen.


1. Am I to understand that regardless of whether interim results are enabled in conjunction with endpointing, Deepgram will send back parts of the finalized transcripts in is_final: true messages even before the threshold silence has been exceeded?

^ this is true - if this were not true, someone might speak a 60 second long run-on sentence, and you might start to wonder why Deepgram isn't returning any results for that full 60 seconds (just as an example)

 2. Does Endpointing rely on pure silence? Or the lack of what it determines to be a voice? I'm trying to figure out why occasionally when endpointing is set to higher ms, it holds back a speech_final: true even when no one is talking and the silence threshold has been exceeded. I've even muted myself on Twilio to test this, and it only sends the final part of the transcript once I begin to talk again. My assumption is that it's because Twilio is a noisier connection so the threshold is not hit if based on pure silence.

^ it triggers when it determines that a voice has stopped (it uses a VAD, not a notion of silence like decibels for example).

> I've even muted myself on Twilio to test this, and it only sends the final part of the transcript once I begin to talk again

this can be tricky - I would have to double check, but when muting yourself on Twilio, Twilio probably sends no audio at all, in which case the audio stream hasn't technically progressed, so this behavior would make sense - this is something I'd like to test, but would describe that behavior. If you have an example request id where this issue occurs, I can track down what's going on.

3. Since I see it hasn't been documented just yet, is this considered a stable feature or are there any concerns with using it in production?

^ this is a great question! I actually use the `UtteranceEnd` feature for all of my conversational AI demos. I'll follow through to make sure it gets documented, but the short story is yes, it is stable. I will note that it works off of silence as detected by a lack of transcribed words, it does not work off of the raw audio (like the VAD-based endpointing algorithm does) - as such, it has worse granularity, but is robust to non-speech noise (white noise, fans, cars, instruments in the background, etc).

Let me know if these answer some of your questions adequately, and always happy to answer more/go into more detail.

**Reply by rrichc** [reply]
>  but when muting yourself on Twilio, Twilio probably sends no audio at all, in which case the audio stream hasn't technically progressed, so this behavior would make sense 

I should clarify I muted on my phone while calling a Twilio number, not sure if that makes a difference between a hardware mute locally vs. directly muting through their API (if one exists). 

> If you have an example request id where this issue occurs, I can track down what's going on.

If I encounter it again I'll make sure to circle back to this post with the request id!

Great answers to my questions, thank you!
